diesel/src/query_builder/insert_statement/mod.rs

```diff
@@ -160,7 +170,11 @@ where
 }
 
 impl<T: QuerySource, U, Op, Ret> InsertStatement<T, U, Op, Ret> {
-    fn new(target: T, records: U, operator: Op, returning: Ret) -> Self {
+    /// Create a new InsertStatement instance
+    #[diesel_derives::__diesel_public_if(
+        feature = "i-implement-a-third-party-backend-and-opt-into-breaking-changes"
+    )]
+    pub(crate) fn new(target: T, records: U, operator: Op, returning: Ret) -> Self {
```

Comment:
The idea behind this macro is to basically generate the following two versions of this function (or any other item which it is applied to). 

```rust

#[cfg(not(feature = "i-implement-a-third-party-backend-and-opt-into-breaking-changes"))]
pub(crate) fn new(target: T, records: U, operator: Op, returning: Ret) ) -> … { …}

#[cfg(feature = "i-implement-a-third-party-backend-and-opt-into-breaking-changes")]
pub fn new(target: T, records: U, operator: Op, returning: Ret) ) -> … { …}

```

It's just a helper that prevents duplicating code when something should be only available behind a specific feature flag. 

I've opted for having the "default" visibility (that one for `not(feature = …)`) on the item itself, because there are cases where you want to use `pub(crate)` or `pub(in crate::pg)` or whatever in that position. For making such items public, the macro always needs to replace whatever visibility is there with just `pub`, which is in my opinion easier than handling that the other way around.

As for the name: I'm happy to rename that to whatever is a better name in a follow up PR.